### PR TITLE
xbutil2: Add support for scan|query|dmatest sub commands

### PR DESCRIPTION
**Pull request by rajkumar-xilinx**
Added changes in windowsXRT to work with scan, query and
dmatest sub commands of xbutil tool.

Resolve review comments for PR# 2499

Signed-off-by: Rajkumar Rampelli <rajkumar@xilinx.com>

### CODE REVIEW

**Comment by rajkumar-xilinx** [comment]
This PR depends on XoclUser driver PR https://github.com/Xilinx/MS_XOCLUSER/pull/13

**Comment by rajkumar-xilinx** [comment]
xbutil scan output on windows:

C:\Rajkumar\repos\27nov_raj_xrt\XRT\build>xbutil --override scan
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
System Configuration
OS Name       : Windows Server 2016 Datacenter Evaluation
Release       : 14393.rs1_release.161220-1747
Version       : 6.3
Machine       : x86_64
Glibc         : N/A
Distribution  : N/A
Now           : Mon Dec  2 17:16:23 2019

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
XRT Information
Version       : 2.4.0
Git Hash      : 2f8d1faf3de55dbde4f9ac1bad29a133c6d4673d
Git Branch    : 2dec_pr
Build Date    : 2019-12-02 16:49:03
XOCL          : N/A
XCLMGMT       : N/A
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[0]: xilinx_u250_xdma_201830_2(ts=1560443408)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
INFO: Found total 1 card(s), 1 are usable.

**Comment by rajkumar-xilinx** [comment]
xbutil query output:
----------------------

C:\Rajkumar\repos\27nov_raj_xrt\XRT\build>xbutil --override query
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
System Configuration
OS Name       : Windows Server 2016 Datacenter Evaluation
Release       : 14393.rs1_release.161220-1747
Version       : 6.3
Machine       : x86_64
Glibc         : N/A
Distribution  : N/A
Now           : Mon Dec  2 17:16:26 2019

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
XRT Information
Version       : 2.4.0
Git Hash      : 2f8d1faf3de55dbde4f9ac1bad29a133c6d4673d
Git Branch    : 2dec_pr
Build Date    : 2019-12-02 16:49:03
XOCL          : N/A
XCLMGMT       : N/A
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compute Unit Status:
[0]: bandwidth1:kernel_1 @0x1800000     (state:TBD)
[1]: bandwidth2:kernel_2 @0x1810000     (state:TBD)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Memory Status:
Tag            Type         Temp(C)       Size (GB)       Mem Usage        BO count
bank0        MEM_DDR4             N/A              16               0 (MB)               0
bank1        MEM_DDR4             N/A              16               0 (MB)               0
bank2        MEM_DDR4             N/A              16               0 (MB)               0
bank3        MEM_DDR4             N/A              16               0 (MB)               0

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Shell                           FPGA                            IDCode
xilinx_u250_xdma_201830_2       xcu250-figd2104-2L-e            N/A
Vendor          Device          SubDevice       SubVendor       SerNum
4334            20485           14              4334            N/A
DDR size (MB)   DDR count       Clock0          Clock1          Clock2
16384           4               -1              -1              -1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by rajkumar-xilinx** [comment]
xbutil dmatest output:
-------------------------

C:\Rajkumar\repos\27nov_raj_xrt\XRT\build>xbutil --override dmatest -b 1024
Total DDR size: 65536 MB
Reporting from mem_topology:
Memory regions:4
Data Validity & DMA Test on bank0:
Buffer Size: 1 MB
Host -> PCIe -> FPGA write bandwidth = 9168.66 MB/s
Host <- PCIe <- FPGA read bandwidth = 9976.45 MB/s
Data Validity & DMA Test on bank1:
Buffer Size: 1 MB
Host -> PCIe -> FPGA write bandwidth = 9248.78 MB/s
Host <- PCIe <- FPGA read bandwidth = 9972.95 MB/s
Data Validity & DMA Test on bank2:
Buffer Size: 1 MB
Host -> PCIe -> FPGA write bandwidth = 9322.16 MB/s
Host <- PCIe <- FPGA read bandwidth = 9977.13 MB/s
Data Validity & DMA Test on bank3:
Buffer Size: 1 MB
Host -> PCIe -> FPGA write bandwidth = 8927.38 MB/s
Host <- PCIe <- FPGA read bandwidth = 10027.8 MB/s

File: src/runtime_src/core/include/xclbin.h [hunk 1]
```diff
@@ -38,6 +38,8 @@
 #ifndef _XCLBIN_H_
 #define _XCLBIN_H_
 
+#define MAX_DDR_BANKS    4
```
Comment (rajkumar-xilinx) [inline src/runtime_src/core/include/xclbin.h]: Need to add XoclUser driver changes in order to remove this macro def here

**Comment by rajkumar-xilinx** [comment]
>>>[Soren] You need to create a config.h file that sets up export and import macros for symbols that are exported out of libxrt_core.dll. Take a look at runtime_src/core/common/config.h which serves the same purpose for the libxrt_coreutil.dll. You could use XRT_CORE_SHIM_EXPORT/IMPORT in place of XRT_CORE_UTIL_EXPORT/IMPORT.
You include config.h in windows/shim.h and in shim.cpp you #define XRT_SHIM_SOURCE before including shim.h.

I will work on this, and will update

**Comment by rajkumar-xilinx** [comment]
>>>[soren] Please update PR to #2503 changes and address the review comments.
Done.

**Comment by rajkumar-xilinx** [comment]
Build error:
[ 67%] Linking CXX executable xbutil2
/usr/bin/ld: CMakeFiles/xbutil2.dir/SubCmdDmaTest.cpp.o: undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
//lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
runtime_src/core/tools/xbutil2/CMakeFiles/xbutil2.dir/build.make:385: recipe for target 'runtime_src/core/tools/xbutil2/xbutil2' failed
make[2]: *** [runtime_src/core/tools/xbutil2/xbutil2] Error 1

File: src/runtime_src/core/include/xclbin.h (hunk 1 above)
Comment (stsoe) [reply src/runtime_src/core/include/xclbin.h]: Maybe so, but this PR cannot be submitted before all changes to xclbin.h have been removed.

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 2]
```diff
@@ -223,4 +223,31 @@ update_SC(const std::string& file) const
   std::cout << "TO-DO: update_SC\n";
 }
 
+unsigned long
+device_linux::
+get_ip_layoutsize(uint64_t _deviceID) const
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/device_linux.cpp]: deviceID is no longer part of the device_* class objects.  The device itself is created from a device index.

File: src/runtime_src/core/pcie/linux/device_linux.h [hunk 3]
```diff
@@ -41,6 +41,10 @@ class device_linux : public device_pcie
   virtual void reset_shell() const;
   virtual void update_shell(const std::string& flashType, const std::string& primary, const std::string& secondary) const;
   virtual void update_SC(const std::string& file) const;
+  virtual void get_ip_layout(uint64_t _deviceID, struct ip_layout **ipLayout, unsigned long size) const;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/device_linux.h]: Remove deviceID per earlier comment

File: src/runtime_src/core/pcie/windows/shim.cpp [hunk 4]
```diff
@@ -1105,3 +1147,157 @@ xclRead(xclDeviceHandle handle, enum xclAddressSpace space,
   auto shim = get_shim_object(handle);
   return shim->read(space,offset,hostbuf,size) ? 0 : size;
 }
+
+void queryDeviceWithQR(uint64_t _deviceID, boost::any & _returnValue,
+	int QR_ID, const std::type_info & _typeInfo, uint64_t statClass)
+{
+  xclDeviceHandle handle = xclOpen((int)_deviceID, 0, XCL_INFO);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/windows/shim.cpp]: A shim level function should work on an already open device handle.   Who ever calls this function should make sure the device is open and pass the handle.  If you are calling this function through a xrt_core::device object, then the device object has already opened the device so all you need is to pass the handle down.

File: src/runtime_src/core/pcie/windows/shim.h [hunk 5]
```diff
@@ -19,6 +19,20 @@
 
 #include "xrt.h"
 #include "core/common/xrt_profiling.h"
+#include "boost/any.hpp"
+#include <boost/property_tree/ptree.hpp>
+
+ // To be simplified
+#include "core/pcie/driver/windows/include/XoclUser_INTF.h"
+
+__declspec(dllexport) void queryDeviceWithQR(uint64_t _deviceID, boost::any & _returnValue, int QR_ID, const std::type_info & _typeInfo, uint64_t statClass);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/windows/shim.h]: A shim level function should work on an already open device handle. Who ever calls this function should make sure the device is open and pass the handle. If you are calling this function through a xrt_core::device object, then the device object has already opened the device so all you need is to pass the handle down.

File: src/runtime_src/core/tools/xbutil2/XBDatabase.cpp [hunk 6]
```diff
@@ -132,8 +165,66 @@ XBDatabase::create_complete_device_tree(boost::property_tree::ptree & _pt)
...
       ptPlatform.add_child("pcie_dma", pt);
     }
+#endif
+
+    {
+	  std::cout << "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~" << std::endl;
+	  //boost::property_tree::ptree pt;
+	  struct ip_layout *ipLayout = NULL;
+	  unsigned long ipSize = device->get_ip_layoutsize(device_id);
+	  ipLayout = (struct ip_layout*)malloc(ipSize);
```
Comment (stsoe) [inline src/runtime_src/core/tools/xbutil2/XBDatabase.cpp]: Memory leak ?

**Comment by rajkumar-xilinx** [comment]
Duplicate of https://github.com/Xilinx/XRT/pull/2505

**Comment by rajkumar-xilinx** [comment]
Resolved few of the comments as part of https://github.com/Xilinx/XRT/pull/2505

**Pull request closed without merging**